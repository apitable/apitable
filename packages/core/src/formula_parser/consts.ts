
// reserved keyword: values are used to rollup the lookup field
export const ROLLUP_KEY_WORDS = 'values';