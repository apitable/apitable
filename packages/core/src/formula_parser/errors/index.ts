
export * from './params_count.error';
export * from './self_ref.error';
export * from './unit.error';