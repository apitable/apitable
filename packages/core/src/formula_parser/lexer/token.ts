/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export enum TokenType {
  Call = 'Call',
  PureValue = 'PureValue',
  Value = 'Value',
  And = 'And',
  Comma = 'Comma',
  Number = 'Number',
  String = 'String',
  Less = 'Less',
  LessEqual = 'LessEqual',
  Greater = 'Greater',
  GreaterEqual = 'GreaterEqual',
  Equal = 'Equal',
  NotEqual = 'NotEqual',
  Add = 'Add',
  Minus = 'Minus',
  Times = 'Times',
  Mod = 'Mod',
  Div = 'Div',
  Concat = 'Concat',
  Or = 'Or',
  Not = 'Not',
  LeftParen = 'LeftParen',
  RightParen = 'RightParen',
  Blank = 'Blank',
  Unknown = 'Unknown',
}

export class Token {
  constructor(readonly type: TokenType, readonly index: number, readonly value: string) {
    // String and Value need to remove quotes and braces
    this.value = value;
  }

  toString() {
    return `${this.type}::${this.value}`;
  }
}
