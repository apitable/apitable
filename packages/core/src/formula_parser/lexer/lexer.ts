/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Token, TokenType } from './token';
import { t, Strings } from 'exports/i18n';
import { isNumber } from 'lodash';
export const EXPR_GRAMMAR: { key: TokenType; exp: RegExp }[] = [
  // The value in the record obtained by the field name constant
  { key: TokenType.Value, exp: /(\{\})|(\{(\\[{}])*[\s\S]*?[^\\]\})/ },
  // string literal
  { key: TokenType.String, exp: /(["”“](([^\\"“”])*?\\["”“])*.*?["”“])|(['‘’](([^\\'‘’])*?\\['‘’])*.*?['‘’])/ },
  // function name or field name constant
  { key: TokenType.Call, exp: /[^0-9.+\-|=*/><()（）!&%'"“”‘’^`~,，\s][^+\-|=*/><()（）!&%'"“”‘’^`~,，\s]*/ },
  // number literal
  { key: TokenType.Number, exp: /[0-9.]+/ },
  // not equal to
  { key: TokenType.NotEqual, exp: /!=/ },
  // and
  { key: TokenType.And, exp: /&&/ },
  // greater or equal to
  { key: TokenType.GreaterEqual, exp: />=/ },
  // less than or equal to
  { key: TokenType.LessEqual, exp: /<=/ },
  // or
  { key: TokenType.Or, exp: /\|\|/ },
  // comma, parameter separator
  { key: TokenType.Comma, exp: /[,，]/ },
  // Not
  { key: TokenType.Not, exp: /!/ },
  // add +
  { key: TokenType.Add, exp: /\+/ },
  // reduce -
  { key: TokenType.Minus, exp: /-/ },
  // take *
  { key: TokenType.Times, exp: /\*/ },
  // remove /
  { key: TokenType.Div, exp: /\// },
  // take remainder %
  { key: TokenType.Mod, exp: /%/ },
  // string concatenation
  { key: TokenType.Concat, exp: /&/ },
  // more than the
  { key: TokenType.Greater, exp: />/ },
  // less than
  { key: TokenType.Less, exp: /</ },
  // equal to
  { key: TokenType.Equal, exp: /=/ },
  // Left parenthesis
  { key: TokenType.LeftParen, exp: /[(（]/ },
  // closing parenthesis
  { key: TokenType.RightParen, exp: /[)）]/ },
  // whitespace characters
  { key: TokenType.Blank, exp: /\s+/ },
  // all other
  { key: TokenType.Unknown, exp: /.+/ },
];
// {} - {midterm test scores}
export interface ILexer {
  expression: string;
  currentTokenIndex: number;

  getNextToken(): Token | null;
  getPrevToken(): Token | null;
  reset(): void;
}
// The lexical analyzer of the formula, according to the incoming expression, exper -> token
export class FormulaExprLexer implements ILexer {
  readonly expression: string;

  readonly fullMatches: Token[];
  readonly matches: Token[];
  readonly errors: Error[] = []; // The lexer does not throw errors directly, but temporarily stores the errors for the caller to handle
  public currentIndex: number;

  constructor(expression: string) {
    this.expression = expression;

    this.fullMatches = this.getFullMatches();
    this.matches = this.filterUselessToken(this.fullMatches);
    this.currentIndex = -1;
  }

  get currentTokenIndex() {
    return this.currentIndex;
  }

  set currentTokenIndex(index: number) {
    if (!isNumber(index)) return;
    this.currentIndex = index;
  }

  getNextToken(): Token | null {
    this.currentIndex++;

    if (this.currentIndex > this.matches.length - 1) {
      return null;
    }

    const token = this.matches[this.currentIndex]!;
    return token;
  }

  getPrevToken(): Token | null {
    this.currentIndex--;

    if (this.currentIndex < 0) {
      return null;
    }

    const token = this.matches[this.currentIndex]!;
    return token;
  }

  reset(): void {
    this.currentIndex = -1;
  }

  private getFullMatches(): Token[] {
    const matched: string[] = this.expression.match(this.pattern()) || [];
    let index = 0;
    return matched.map((str, i) => {
      const token = this.tokenizer(index, str, matched[i + 1]);
      index += token.value.length;
      return token;
    });
  }

  private filterUselessToken(tokens: Token[]) {
    return tokens.filter(token => {
      if (token.type === TokenType.Unknown) {
        this.errors.push(
          new Error(
            t(Strings.function_err_unrecognized_operator, {
              token: token.value,
            }),
          ),
        );
      }
      return token.type !== TokenType.Blank && token.type !== TokenType.Unknown;
    });
  }

  private tokenizer(index: number, str: string, nextStr?: string): Token {
    for (let i = 0; i < EXPR_GRAMMAR.length; i++) {
      const key = EXPR_GRAMMAR[i]!.key;
      const regex = EXPR_GRAMMAR[i]!.exp;
      const type = key;

      if (regex.test(str)) {
        // When the tokenType matches Call, it is necessary to judge whether there is a left bracket, if not,
        // it means a pureValue without curly brackets
        if (type === TokenType.Call) {
          if (!nextStr || this.tokenizer(index, nextStr).type !== TokenType.LeftParen) {
            return new Token(TokenType.PureValue, index, str);
          }
        }
        return new Token(type, index, str);
      }
    }

    throw new Error(`Unexpected token: ${str}`);
  }

  private pattern(): RegExp {
    const pattern: string = EXPR_GRAMMAR.map(g => `(${g.exp.source})`).join('|');

    return new RegExp(pattern, 'g');
  }
}
