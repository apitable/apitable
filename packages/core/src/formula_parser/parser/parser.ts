/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import type { ILexer } from '../lexer/lexer';
import { Token, TokenType } from '../lexer/token';
import {
  AstNode,
  BinaryOperatorNode,
  UnaryOperatorNode,
  ValueOperandNode,
  CallOperandNode,
  NumberOperandNode,
  StringOperandNode,
  PureValueOperandNode,
} from './ast';
import type { IFieldMap, IReduxState } from 'exports/store';
import { Functions } from 'formula_parser/functions';
import type { IField } from 'types';
import { t, Strings } from 'exports/i18n';
import { SelfRefError } from 'formula_parser/errors/self_ref.error';

/**
  * operator precedence
  * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence
  * In the array below, the priority is sorted from high to low. The same line has the same priority.
  */
const PriorityMap = new Map<TokenType, number>();
[
  [TokenType.Times, TokenType.Div, TokenType.Mod],
  [TokenType.Add, TokenType.Minus],
  [TokenType.Greater, TokenType.GreaterEqual, TokenType.Less, TokenType.LessEqual],
  [TokenType.Equal, TokenType.NotEqual],
  [TokenType.And],
  [TokenType.Or],
  [TokenType.Concat],
].forEach((arr, index) => arr.forEach(type => PriorityMap.set(type, index)));

export class FormulaExprParser {
  private currentToken: Token | null;

  constructor(public lexer: ILexer, public context: { state: IReduxState, field: IField, fieldMap: IFieldMap }) {
    this.currentToken = this.lexer.getNextToken();
  }

  parse(): AstNode {
    const node = this.expr();
    if (this.currentToken != null) {
      console.error(this.currentToken);
      throw new Error(t(Strings.function_err_unrecognized_char, {
        value: this.currentToken.value,
      }));
    }
    return node;
  }

  private next(type: TokenType): Token | null {
    if (this.currentToken == null) {
      return null;
    }
    if (this.currentToken.type === type) {
      this.currentToken = this.lexer.getNextToken();
    } else {
      switch (type) {
        case TokenType.LeftParen:
          throw new SyntaxError(t(Strings.function_err_no_left_bracket));
        case TokenType.RightParen:
          throw new SyntaxError(t(Strings.function_err_end_of_right_bracket));
      }
      throw new SyntaxError(t(Strings.function_err_unable_parse_char, {
        value: this.currentToken.value,
      }));
    }
    return this.currentToken;
  }

  private factor(): AstNode {
    // factor : VALUE | LEFT_PAREN expr RIGHT_PAREN | NOT expr

    const token = this.currentToken;

    if (!token) {
      throw new Error(t(Strings.function_err_wrong_function_suffix));
    }

    switch (token.type) {
      // field variable: {value}
      case TokenType.Value: {
        this.next(TokenType.Value);
        const tokenValue = token.value.slice(1, -1);
        if (tokenValue === this.context.field.id) {
          throw new SelfRefError();
        }
        return new ValueOperandNode(token, this.context);
      }

      // field variable: value (without curly braces)
      case TokenType.PureValue: {
        this.next(TokenType.PureValue);
        const tokenValue = token.value;
        if (tokenValue === this.context.field.id) {
          throw new SelfRefError();
        }
        return new PureValueOperandNode(token, this.context, this.context.field);
      }

      // Preset functions: Sum/Average ...
      case TokenType.Call: {
        this.next(TokenType.Call);
        const node = new CallOperandNode(token);
        const FuncClass = Functions.get(node.value.toUpperCase());
        if (!FuncClass) {
          throw new TypeError(t(Strings.function_err_not_definition, {
            name: node.value,
          }));
        }

        this.next(TokenType.LeftParen);

        if (!this.currentToken) {
          throw new Error(t(Strings.function_err_end_of_right_bracket));
        }

        while (this.currentToken.type !== TokenType.RightParen) {
          node.params.push(this.expr());
          if (!this.currentToken) {
            throw new Error(t(Strings.function_err_end_of_right_bracket));
          }
          // Exclude multiple parameters without commas
          if (this.currentToken.type !== TokenType.Comma) {
            break;
          }
        }

        const valueType = FuncClass.func.getReturnType(node.params);
        node.valueType = valueType;

        this.next(TokenType.RightParen);
        return node;
      }

      // number: 123.333
      case TokenType.Number: {
        this.next(TokenType.Number);
        return new NumberOperandNode(token);
      }

      // string: 'xyz'
      case TokenType.String: {
        this.next(TokenType.String);
        return new StringOperandNode(token);
      }

      // Left parenthesis: '('
      case TokenType.LeftParen: {
        this.next(TokenType.LeftParen);
        const node: AstNode = this.expr();
        this.next(TokenType.RightParen);
        return node;
      }

      // Negate sign (unary arithmetic sign): '!'
      case TokenType.Not: {
        this.next(TokenType.Not);
        const node: AstNode = this.factor();
        return new UnaryOperatorNode(node, token);
      }

      // + sign (unary arithmetic sign): '+'
      case TokenType.Add: {
        this.next(TokenType.Add);
        const node: AstNode = this.factor();
        return new UnaryOperatorNode(node, token);
      }

      // -sign (unary arithmetic sign): '-'
      case TokenType.Minus: {
        this.next(TokenType.Minus);
        const node: AstNode = this.factor();
        return new UnaryOperatorNode(node, token);
      }

      case TokenType.Comma: {
        this.next(TokenType.Comma);
        const node: AstNode = this.expr();
        return node;
      }

      case TokenType.Blank: {
        this.next(TokenType.Blank);
        return this.factor();
      }

      default:
        throw new Error(t(Strings.function_err_unknown_operator, {
          type: token.value,
        }));
    }
  }

  private expr(inner?: boolean): AstNode {
    // expr   : factor ((&& | ||) factor)*
    // factor : Number | String | Call | VALUE | LEFT_PAREN expr RIGHT_PAREN | NOT expr

    let node: AstNode = this.factor();

    while (this.currentToken &&
      [
        TokenType.And, TokenType.Or, TokenType.Add, TokenType.Times, TokenType.Div, TokenType.Minus,
        TokenType.Mod, TokenType.Concat, TokenType.Equal, TokenType.NotEqual, TokenType.Greater, TokenType.GreaterEqual,
        TokenType.Less, TokenType.LessEqual,
      ].includes(this.currentToken.type)
    ) {
      const token: Token = this.currentToken;
      this.next(token.type);
      if (!this.currentToken) {
        throw new Error(t(Strings.function_err_wrong_function_suffix));
      }
      let right: AstNode | undefined;
      let nextToken: Token | null = null;
      const currentToken = this.currentToken;
      const currentTokenIndex = this.lexer.currentTokenIndex;
      /**
        * Take a step forward, get the token and go back
        * 
        * 1. If you encounter a function or left parenthesis, go forward to test the entire function or parenthesis content, 
        * get the following operator and then fall back
        * 
        * 2. If it is not a function, just try a token forward, get the operator and then fall back
        */
      if ([TokenType.Call, TokenType.LeftParen].includes(currentToken.type)) {
        this.factor();
        nextToken = this.currentToken;
        this.currentToken = currentToken;
        this.lexer.currentTokenIndex = currentTokenIndex;
      } else {
        nextToken = this.lexer.getNextToken();
        this.lexer.getPrevToken();
      }

      const currentOpIndex = PriorityMap.get(token.type);
      if (nextToken) {
        const nextOpIndex = PriorityMap.get(nextToken.type);
      
        if (currentOpIndex != null && nextOpIndex != null && nextOpIndex < currentOpIndex) {
          right = this.expr(true);
        }
        // When operators with different priorities are encountered in the loop, the recursion must be exited;
        if (inner && currentOpIndex != null && nextOpIndex != null && nextOpIndex > currentOpIndex) {
          return new BinaryOperatorNode(node, token, right || this.factor());
        }
      }
      node = new BinaryOperatorNode(node, token, right || this.factor());
    }
    // 1 + (1 + 3) * 2
    return node;
  }
}
