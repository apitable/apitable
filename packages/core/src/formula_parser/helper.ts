/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { handleNullArray } from 'model';
import { FieldType } from 'types';
import { IFormulaParam } from './functions/basic';
import { AstNodeType, ValueOperandNode } from './parser';
import { produce } from 'immer';

/**
 * lookup cv now leaves null by default. 
 * When the formula refers to the lookup field as a parameter, some functions need to convert the null value to make it as expected.
 */
export const handleLookupNullValue = (params: IFormulaParam<any>[]) => {
  return produce(params, draftParams => {
    draftParams.forEach(param => {
      if (param.node.name === AstNodeType.ValueOperandNode
        && (param.node as ValueOperandNode).field.type === FieldType.LookUp
        && handleNullArray(param.value) == null
      ) {
        param.value = null;
      }
    });
  });
};