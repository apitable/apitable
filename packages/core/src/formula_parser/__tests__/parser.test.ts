import { Strings, t } from 'exports/i18n';
import { IReduxState } from 'exports/store';
import { FormulaExprLexer } from 'formula_parser/lexer';
import { AstNode, FormulaExprParser } from 'formula_parser/parser';
import { DateFormat, FieldType, TimeFormat } from 'types';

const mockContext = (expr: string): ConstructorParameters<typeof FormulaExprParser>[1] => ({
  state: {} as IReduxState,
  field: {
    id: 'fld4',
    name: 'Field 4',
    type: FieldType.Formula,
    property: {
      datasheetId: 'dst1',
      expression: expr,
    },
  },
  fieldMap: {
    fld1: {
      id: 'fld1',
      name: 'Field 1',
      type: FieldType.SingleText,
      property: {},
    },
    fld2: {
      id: 'fld2',
      name: 'Field 2',
      type: FieldType.Number,
      property: {
        precision: 0,
      },
    },
    fld3: {
      id: 'fld3',
      name: 'Field 3',
      type: FieldType.DateTime,
      property: {
        dateFormat: DateFormat['YYYY-MM'],
        timeFormat: TimeFormat['HH:mm'],
        includeTime: false,
        autoFill: false,
      },
    },
    fld4: {
      id: 'fld4',
      name: 'Field 4',
      type: FieldType.Formula,
      property: {
        datasheetId: 'dst1',
        expression: expr,
      },
    },
  },
});

const parse = (input: string): AstNode => {
  const lexer = new FormulaExprLexer(input);
  const parser = new FormulaExprParser(lexer, mockContext(input));
  return parser.parse();
};

describe('function call', () => {
  test('nullary function', () => {
    const ast = parse('record_id()');
    expect(ast).toMatchSnapshot();
  });

  test('unary function', () => {
    const ast = parse("inT('babe')");
    expect(ast).toMatchSnapshot();
  });

  test('3-arity function', () => {
    const ast = parse('Max(31, -77, {fld3})');
    expect(ast).toMatchSnapshot();
  });

  test('nested function', () => {
    const ast = parse("Max(rOuNd(31+true()), value('')&'b', --0)");
    expect(ast).toMatchSnapshot();
  });

  test('missing right parenthesis', () => {
    expect.assertions(1);
    try {
      parse('max(1,7');
    } catch (e) {
      expect(e).toStrictEqual(new Error(t(Strings.function_err_end_of_right_bracket)));
    }
  });

  test('redundant comma', () => {
    expect.assertions(1);
    try {
      parse('max(1,7,)');
    } catch (e) {
      expect(e).toStrictEqual(new Error(t(Strings.function_err_unknown_operator, { type: ',' })));
    }
  });
});

describe('field', () => {
  test('with curly braces', () => {
    const ast = parse('  {fld2}');
    expect(ast).toMatchSnapshot();
  });

  test('without curly braces', () => {
    const ast = parse('fld2  ');
    expect(ast).toMatchSnapshot();
  });

  test('two consecutive fields', () => {
    expect.assertions(1);
    try {
      parse('fld2  {fld4} ');
    } catch (e) {
      expect(e).toStrictEqual(new Error(t(Strings.function_err_unrecognized_char, { value: '{fld4}' })));
    }
  });
});

describe('unary operator', () => {
  test('not', () => {
    const ast = parse('  ! "foo"');
    expect(ast).toMatchSnapshot();
  });

  test('minus', () => {
    const ast = parse('  - "foo"');
    expect(ast).toMatchSnapshot();
  });

  test('plus', () => {
    const ast = parse('  + "foo"');
    expect(ast).toMatchSnapshot();
  });

  test('nested', () => {
    const ast = parse('  +-!( +--"foo")');
    expect(ast).toMatchSnapshot();
  });
});