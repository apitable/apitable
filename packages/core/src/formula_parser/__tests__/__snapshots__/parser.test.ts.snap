// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field with curly braces 1`] = `
ValueOperandNode {
  "context": {
    "field": {
      "id": "fld4",
      "name": "Field 4",
      "property": {
        "datasheetId": "dst1",
        "expression": "  {fld2}",
      },
      "type": 16,
    },
    "fieldMap": {
      "fld1": {
        "id": "fld1",
        "name": "Field 1",
        "property": {},
        "type": 19,
      },
      "fld2": {
        "id": "fld2",
        "name": "Field 2",
        "property": {
          "precision": 0,
        },
        "type": 2,
      },
      "fld3": {
        "id": "fld3",
        "name": "Field 3",
        "property": {
          "autoFill": false,
          "dateFormat": 3,
          "includeTime": false,
          "timeFormat": 0,
        },
        "type": 5,
      },
      "fld4": {
        "id": "fld4",
        "name": "Field 4",
        "property": {
          "datasheetId": "dst1",
          "expression": "  {fld2}",
        },
        "type": 16,
      },
    },
    "state": {},
  },
  "field": {
    "id": "fld2",
    "name": "Field 2",
    "property": {
      "precision": 0,
    },
    "type": 2,
  },
  "innerValueType": "String",
  "name": "ValueOperandNode",
  "token": Token {
    "index": 2,
    "type": "Value",
    "value": "{fld2}",
  },
  "value": "{fld2}",
  "valueType": "Number",
}
`;

exports[`field without curly braces 1`] = `
PureValueOperandNode {
  "context": {
    "field": {
      "id": "fld4",
      "name": "Field 4",
      "property": {
        "datasheetId": "dst1",
        "expression": "fld2  ",
      },
      "type": 16,
    },
    "fieldMap": {
      "fld1": {
        "id": "fld1",
        "name": "Field 1",
        "property": {},
        "type": 19,
      },
      "fld2": {
        "id": "fld2",
        "name": "Field 2",
        "property": {
          "precision": 0,
        },
        "type": 2,
      },
      "fld3": {
        "id": "fld3",
        "name": "Field 3",
        "property": {
          "autoFill": false,
          "dateFormat": 3,
          "includeTime": false,
          "timeFormat": 0,
        },
        "type": 5,
      },
      "fld4": {
        "id": "fld4",
        "name": "Field 4",
        "property": {
          "datasheetId": "dst1",
          "expression": "fld2  ",
        },
        "type": 16,
      },
    },
    "state": {},
  },
  "field": {
    "id": "fld2",
    "name": "Field 2",
    "property": {
      "precision": 0,
    },
    "type": 2,
  },
  "innerValueType": "String",
  "name": "PureValueOperandNode",
  "token": Token {
    "index": 0,
    "type": "PureValue",
    "value": "fld2",
  },
  "value": "fld2",
  "valueType": "Number",
}
`;

exports[`function call 3-arity function 1`] = `
CallOperandNode {
  "name": "CallOperandNode",
  "params": [
    NumberOperandNode {
      "name": "NumberOperandNode",
      "token": Token {
        "index": 4,
        "type": "Number",
        "value": "31",
      },
      "value": "31",
      "valueType": "Number",
    },
    UnaryOperatorNode {
      "child": NumberOperandNode {
        "name": "NumberOperandNode",
        "token": Token {
          "index": 9,
          "type": "Number",
          "value": "77",
        },
        "value": "77",
        "valueType": "Number",
      },
      "name": "UnaryOperatorNode",
      "token": Token {
        "index": 8,
        "type": "Minus",
        "value": "-",
      },
      "valueType": "Number",
    },
    ValueOperandNode {
      "context": {
        "field": {
          "id": "fld4",
          "name": "Field 4",
          "property": {
            "datasheetId": "dst1",
            "expression": "Max(31, -77, {fld3})",
          },
          "type": 16,
        },
        "fieldMap": {
          "fld1": {
            "id": "fld1",
            "name": "Field 1",
            "property": {},
            "type": 19,
          },
          "fld2": {
            "id": "fld2",
            "name": "Field 2",
            "property": {
              "precision": 0,
            },
            "type": 2,
          },
          "fld3": {
            "id": "fld3",
            "name": "Field 3",
            "property": {
              "autoFill": false,
              "dateFormat": 3,
              "includeTime": false,
              "timeFormat": 0,
            },
            "type": 5,
          },
          "fld4": {
            "id": "fld4",
            "name": "Field 4",
            "property": {
              "datasheetId": "dst1",
              "expression": "Max(31, -77, {fld3})",
            },
            "type": 16,
          },
        },
        "state": {},
      },
      "field": {
        "id": "fld3",
        "name": "Field 3",
        "property": {
          "autoFill": false,
          "dateFormat": 3,
          "includeTime": false,
          "timeFormat": 0,
        },
        "type": 5,
      },
      "innerValueType": "String",
      "name": "ValueOperandNode",
      "token": Token {
        "index": 13,
        "type": "Value",
        "value": "{fld3}",
      },
      "value": "{fld3}",
      "valueType": "DateTime",
    },
  ],
  "token": Token {
    "index": 0,
    "type": "Call",
    "value": "Max",
  },
  "value": "Max",
  "valueType": "Number",
}
`;

exports[`function call nested function 1`] = `
CallOperandNode {
  "name": "CallOperandNode",
  "params": [
    CallOperandNode {
      "name": "CallOperandNode",
      "params": [
        BinaryOperatorNode {
          "left": NumberOperandNode {
            "name": "NumberOperandNode",
            "token": Token {
              "index": 10,
              "type": "Number",
              "value": "31",
            },
            "value": "31",
            "valueType": "Number",
          },
          "name": "BinaryOperatorNode",
          "right": CallOperandNode {
            "name": "CallOperandNode",
            "params": [],
            "token": Token {
              "index": 13,
              "type": "Call",
              "value": "true",
            },
            "value": "true",
            "valueType": "Boolean",
          },
          "token": Token {
            "index": 12,
            "type": "Add",
            "value": "+",
          },
          "valueType": "String",
        },
      ],
      "token": Token {
        "index": 4,
        "type": "Call",
        "value": "rOuNd",
      },
      "value": "rOuNd",
      "valueType": "Number",
    },
    BinaryOperatorNode {
      "left": CallOperandNode {
        "name": "CallOperandNode",
        "params": [
          StringOperandNode {
            "name": "StringOperandNode",
            "token": Token {
              "index": 28,
              "type": "String",
              "value": "''",
            },
            "value": "''",
            "valueType": "String",
          },
        ],
        "token": Token {
          "index": 22,
          "type": "Call",
          "value": "value",
        },
        "value": "value",
        "valueType": "Number",
      },
      "name": "BinaryOperatorNode",
      "right": StringOperandNode {
        "name": "StringOperandNode",
        "token": Token {
          "index": 32,
          "type": "String",
          "value": "'b'",
        },
        "value": "'b'",
        "valueType": "String",
      },
      "token": Token {
        "index": 31,
        "type": "Concat",
        "value": "&",
      },
      "valueType": "String",
    },
    UnaryOperatorNode {
      "child": UnaryOperatorNode {
        "child": NumberOperandNode {
          "name": "NumberOperandNode",
          "token": Token {
            "index": 39,
            "type": "Number",
            "value": "0",
          },
          "value": "0",
          "valueType": "Number",
        },
        "name": "UnaryOperatorNode",
        "token": Token {
          "index": 38,
          "type": "Minus",
          "value": "-",
        },
        "valueType": "Number",
      },
      "name": "UnaryOperatorNode",
      "token": Token {
        "index": 37,
        "type": "Minus",
        "value": "-",
      },
      "valueType": "Number",
    },
  ],
  "token": Token {
    "index": 0,
    "type": "Call",
    "value": "Max",
  },
  "value": "Max",
  "valueType": "Number",
}
`;

exports[`function call nullary function 1`] = `
CallOperandNode {
  "name": "CallOperandNode",
  "params": [],
  "token": Token {
    "index": 0,
    "type": "Call",
    "value": "record_id",
  },
  "value": "record_id",
  "valueType": "String",
}
`;

exports[`function call unary function 1`] = `
CallOperandNode {
  "name": "CallOperandNode",
  "params": [
    StringOperandNode {
      "name": "StringOperandNode",
      "token": Token {
        "index": 4,
        "type": "String",
        "value": "'babe'",
      },
      "value": "'babe'",
      "valueType": "String",
    },
  ],
  "token": Token {
    "index": 0,
    "type": "Call",
    "value": "inT",
  },
  "value": "inT",
  "valueType": "Number",
}
`;

exports[`unary operator minus 1`] = `
UnaryOperatorNode {
  "child": StringOperandNode {
    "name": "StringOperandNode",
    "token": Token {
      "index": 4,
      "type": "String",
      "value": ""foo"",
    },
    "value": ""foo"",
    "valueType": "String",
  },
  "name": "UnaryOperatorNode",
  "token": Token {
    "index": 2,
    "type": "Minus",
    "value": "-",
  },
  "valueType": "Number",
}
`;

exports[`unary operator nested 1`] = `
UnaryOperatorNode {
  "child": UnaryOperatorNode {
    "child": UnaryOperatorNode {
      "child": UnaryOperatorNode {
        "child": UnaryOperatorNode {
          "child": UnaryOperatorNode {
            "child": StringOperandNode {
              "name": "StringOperandNode",
              "token": Token {
                "index": 10,
                "type": "String",
                "value": ""foo"",
              },
              "value": ""foo"",
              "valueType": "String",
            },
            "name": "UnaryOperatorNode",
            "token": Token {
              "index": 9,
              "type": "Minus",
              "value": "-",
            },
            "valueType": "Number",
          },
          "name": "UnaryOperatorNode",
          "token": Token {
            "index": 8,
            "type": "Minus",
            "value": "-",
          },
          "valueType": "Number",
        },
        "name": "UnaryOperatorNode",
        "token": Token {
          "index": 7,
          "type": "Add",
          "value": "+",
        },
        "valueType": "Number",
      },
      "name": "UnaryOperatorNode",
      "token": Token {
        "index": 4,
        "type": "Not",
        "value": "!",
      },
      "valueType": "Boolean",
    },
    "name": "UnaryOperatorNode",
    "token": Token {
      "index": 3,
      "type": "Minus",
      "value": "-",
    },
    "valueType": "Number",
  },
  "name": "UnaryOperatorNode",
  "token": Token {
    "index": 2,
    "type": "Add",
    "value": "+",
  },
  "valueType": "Number",
}
`;

exports[`unary operator not 1`] = `
UnaryOperatorNode {
  "child": StringOperandNode {
    "name": "StringOperandNode",
    "token": Token {
      "index": 4,
      "type": "String",
      "value": ""foo"",
    },
    "value": ""foo"",
    "valueType": "String",
  },
  "name": "UnaryOperatorNode",
  "token": Token {
    "index": 2,
    "type": "Not",
    "value": "!",
  },
  "valueType": "Boolean",
}
`;

exports[`unary operator plus 1`] = `
UnaryOperatorNode {
  "child": StringOperandNode {
    "name": "StringOperandNode",
    "token": Token {
      "index": 4,
      "type": "String",
      "value": ""foo"",
    },
    "value": ""foo"",
    "valueType": "String",
  },
  "name": "UnaryOperatorNode",
  "token": Token {
    "index": 2,
    "type": "Add",
    "value": "+",
  },
  "valueType": "String",
}
`;
